From 8d20e4e835d152f5a27e309d3ab2f074b21e5a83 Mon Sep 17 00:00:00 2001
From: Dylan Hung <dylan_hung@aspeedtech.com>
Date: Fri, 27 Oct 2023 17:42:22 +0800
Subject: [PATCH] feat(ast2700): add SLI calibration

The AST2700 SoC embeds two dies, the CPU-die and IO-die, and uses a
serial bus (SLI) to connect these two dies. This commit speeds up the
upstream direction (from IO-die to CPU-die) to 500MHz and calibrates the
RX tap delay.

v2: Increase the delay time in the calibration loop
v3: Add SLIM per-PAD delay calibration
v4: Use read-modify-write for IOD SLI registers
v5: Add SLI 250MHz support
v6: Switch SLI US clock frequency to 250MHz
v7:
- Refactor the message print format
- load the default delay if no pass window found
v8:
- Clear bus when clearing TX/RX queues
- Clear interrupt status before exiting sli_init.
v9: Disable printf in release mode

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
---
 plat/aspeed/ast2700/include/platform_reg.h |  64 +++
 plat/aspeed/ast2700/plat_bl31_setup.c      | 445 +++++++++++++++++++++
 plat/aspeed/ast2700/platform.mk            |   2 +
 3 files changed, 511 insertions(+)

diff --git a/plat/aspeed/ast2700/include/platform_reg.h b/plat/aspeed/ast2700/include/platform_reg.h
index 7f268654cfca..92fb1a287a69 100644
--- a/plat/aspeed/ast2700/include/platform_reg.h
+++ b/plat/aspeed/ast2700/include/platform_reg.h
@@ -24,4 +24,68 @@
 #define SCU_CPU_SMP_EP2	(SCU_CPU_BASE + 0x790)
 #define SCU_CPU_SMP_EP3	(SCU_CPU_BASE + 0x798)
 
+/* CPU-die SLI */
+#define SLI_CPU_BASE			0x12c17000
+#define SLIM_CPU_BASE			(SLI_CPU_BASE + 0x000)
+#define SLIH_CPU_BASE			(SLI_CPU_BASE + 0x200)
+#define SLIV_CPU_BASE			(SLI_CPU_BASE + 0x400)
+
+/* IO-die SLI */
+#define SLI_IOD_BASE			0x14c1e000
+#define SLIM_IOD_BASE			(SLI_IOD_BASE + 0x000)
+#define SLIH_IOD_BASE			(SLI_IOD_BASE + 0x200)
+#define SLIV_IOD_BASE			(SLI_IOD_BASE + 0x400)
+
+#define SLI_CTRL_I			0x00
+#define   SLIV_RAW_MODE			BIT(15)
+#define   SLI_TX_MODE			BIT(14)
+#define   SLI_RX_PHY_LAH_SEL_REV	BIT(13)
+#define   SLI_RX_PHY_LAH_SEL_NEG	BIT(12)
+#define   SLI_AUTO_SEND_TRN_OFF		BIT(8)
+#define   SLI_CLEAR_BUS			BIT(6)
+#define   SLI_TRANS_EN			BIT(5)
+#define   SLI_CLEAR_RX			BIT(2)
+#define   SLI_CLEAR_TX			BIT(1)
+#define   SLI_RESET_TRIGGER		BIT(0)
+#define SLI_CTRL_II			0x04
+#define   SLIV_TX_ENT_SUSPEND		GENMASK(15, 14)
+#define SLI_CTRL_III			0x08
+#define   SLI_PHYCLK_SEL		GENMASK(27, 24)
+#define     SLI_PHYCLK_800M		0x1
+#define     SLI_PHYCLK_400M		0x2
+#define     SLI_PHYCLK_200M		0x3
+#define     SLI_PHYCLK_1G		0x5
+#define     SLI_PHYCLK_500M		0x6
+#define     SLI_PHYCLK_250M		0x7
+#define   SLIH_PAD_DLY_TX1		GENMASK(23, 18)
+#define   SLIH_PAD_DLY_TX0		GENMASK(17, 12)
+#define   SLIH_PAD_DLY_RX1		GENMASK(11, 6)
+#define   SLIH_PAD_DLY_RX0		GENMASK(5, 0)
+#define   SLIM_PAD_DLY_RX3		GENMASK(23, 18)
+#define   SLIM_PAD_DLY_RX2		GENMASK(17, 12)
+#define   SLIM_PAD_DLY_RX1		GENMASK(11, 6)
+#define   SLIM_PAD_DLY_RX0		GENMASK(5, 0)
+#define   SLIV_PAD_DLY_TX1		GENMASK(23, 18)
+#define   SLIV_PAD_DLY_TX0		GENMASK(17, 12)
+#define   SLIV_PAD_DLY_RX1		GENMASK(11, 6)
+#define   SLIV_PAD_DLY_RX0		GENMASK(5, 0)
+
+#define SLI_INTR_STATUS			0x14
+#define   SLI_INTR_RX_SYNC		BIT(15)
+#define   SLI_INTR_RX_ERR		BIT(13)
+#define   SLI_INTR_RX_NACK		BIT(12)
+#define   SLI_INTR_RX_TRAIN_PKT		BIT(10)
+#define   SLI_INTR_RX_DISCONN		BIT(6)
+#define   SLI_INTR_TX_SUSPEND		BIT(4)
+#define   SLI_INTR_TX_TRAIN		BIT(3)
+#define   SLI_INTR_TX_IDLE		BIT(2)
+#define   SLI_INTR_RX_SUSPEND		BIT(1)
+#define   SLI_INTR_RX_IDLE		BIT(0)
+#define   SLI_INTR_RX_ERRORS                                                     \
+	  (SLI_INTR_RX_ERR | SLI_INTR_RX_NACK | SLI_INTR_RX_DISCONN)
+
+#define SLI_TARGET_PHYCLK		SLI_PHYCLK_250M
+#define SLIV_DEFAULT_DELAY		5
+#define SLIH_DEFAULT_DELAY		8
+#define SLIM_DEFAULT_DELAY		8
 #endif /* PLATFORM_REG_H */
diff --git a/plat/aspeed/ast2700/plat_bl31_setup.c b/plat/aspeed/ast2700/plat_bl31_setup.c
index 92a48ff805b4..a57913c051f0 100644
--- a/plat/aspeed/ast2700/plat_bl31_setup.c
+++ b/plat/aspeed/ast2700/plat_bl31_setup.c
@@ -10,11 +10,24 @@
 #include <drivers/arm/gicv3.h>
 #include <drivers/console.h>
 #include <drivers/ti/uart/uart_16550.h>
+#include <drivers/generic_delay_timer.h>
+#include <drivers/delay_timer.h>
 #include <lib/mmio.h>
 #include <lib/xlat_tables/xlat_tables_v2.h>
 #include <plat/common/platform.h>
 #include <platform_def.h>
 
+#define __bf_shf(x) (__builtin_ffsll(x) - 1)
+#define FIELD_PREP(_mask, _val)						\
+	({								\
+		((typeof(_mask))(_val) << __bf_shf(_mask)) & (_mask);	\
+	})
+
+#define FIELD_GET(_mask, _reg)						\
+	({								\
+		(typeof(_mask))(((_reg) & (_mask)) >> __bf_shf(_mask));	\
+	})
+
 static console_t console;
 
 static entry_point_info_t bl32_ep_info;
@@ -45,6 +58,10 @@ static const mmap_region_t plat_mmap[] = {
 			MT_DEVICE | MT_RW | MT_SECURE),
 	MAP_REGION_FLAT(SCU_CPU_BASE, PAGE_SIZE,
 			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(SLI_CPU_BASE, PAGE_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
+	MAP_REGION_FLAT(SLI_IOD_BASE, PAGE_SIZE,
+			MT_DEVICE | MT_RW | MT_SECURE),
 	{ 0 }
 };
 
@@ -92,8 +109,436 @@ void bl31_plat_arch_setup(void)
 	enable_mmu_el3(0);
 }
 
+static int sli_wait(uint32_t base, uint32_t mask)
+{
+	uint32_t value;
+
+	mmio_write_32(base + SLI_INTR_STATUS, 0xffffffff);
+
+	do {
+		value = mmio_read_32(base + SLI_INTR_STATUS);
+		if (value & SLI_INTR_RX_ERRORS)
+			return -1;
+	} while ((value & mask) != mask);
+
+	return 0;
+}
+
+static int sli_wait_suspend(uint32_t base)
+{
+	return sli_wait(base, SLI_INTR_TX_SUSPEND | SLI_INTR_RX_SUSPEND);
+}
+
+static int is_sli_suspend(uint32_t base)
+{
+	uint32_t value;
+	uint32_t suspend = SLI_INTR_TX_SUSPEND | SLI_INTR_RX_SUSPEND;
+
+	value = mmio_read_32(base + SLI_INTR_STATUS);
+	if (value & SLI_INTR_RX_ERRORS)
+		return -1;
+	else if ((value & suspend) == suspend)
+		return 1;
+	else
+		return 0;
+}
+
+static void sli_calibrate_ahb_delay(int config)
+{
+	uint32_t reg;
+	int d0, d1, d0_start, d0_end;
+	int win[32][2];
+	int i, latch_sel = 0;
+
+	if (config)
+		latch_sel = SLI_RX_PHY_LAH_SEL_NEG;
+
+	for (i = 0; i < 32; i++) {
+		win[i][0] = -1;
+		win[i][1] = -1;
+	}
+
+	/* Calibrate CPU-SLIH US pad delay */
+	for (d0 = 0; d0 < 16; d0++) {
+		for (d1 = 0; d1 < 16; d1++) {
+			/* Adjust pad delay */
+			reg = mmio_read_32(SLIH_CPU_BASE + SLI_CTRL_III);
+			reg &= ~(SLIH_PAD_DLY_RX1 | SLIH_PAD_DLY_RX0);
+			reg |= FIELD_PREP(SLIH_PAD_DLY_RX1, d1) |
+			       FIELD_PREP(SLIH_PAD_DLY_RX0, d0);
+			mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_III, reg);
+			udelay(1);
+
+			/*
+			 * CPU SLIH
+			 * clr bit[8] -> enable training,
+			 * clr bit[12] -> disable LAH_SEL_NEG
+			 * set bit[2] -> clear RX
+			 */
+			mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_I,
+				      latch_sel | SLI_TRANS_EN | SLI_CLEAR_RX |
+					      SLI_CLEAR_BUS);
+			udelay(100);
+
+			/* check interrupt status */
+			mmio_write_32(SLIH_CPU_BASE + SLI_INTR_STATUS,
+				      0xffffffff);
+			if (is_sli_suspend(SLIH_CPU_BASE)) {
+				if (win[d0][0] == -1)
+					win[d0][0] = d1;
+
+				win[d0][1] = d1;
+			} else if (win[d0][1] != -1) {
+				break;
+			}
+		}
+	}
+
+	d0_start = -1;
+	d0_end = -1;
+	for (i = 0; i < 32; i++) {
+		if (win[i][0] != -1) {
+			if (d0_start == -1)
+				d0_start = i;
+
+			d0_end = i;
+		} else if (d0_end != -1) {
+			break;
+		}
+	}
+
+	if (d0_start == -1) {
+		d0 = SLIH_DEFAULT_DELAY;
+		d1 = SLIH_DEFAULT_DELAY;
+	} else {
+		d0 = (d0_start + d0_end) >> 1;
+		d1 = (win[d0][0] + win[d0][1]) >> 1;
+	}
+
+	/* load the calibrated delay values */
+	mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_I,
+		      latch_sel | SLI_AUTO_SEND_TRN_OFF | SLI_TRANS_EN);
+
+	reg = mmio_read_32(SLIH_CPU_BASE + SLI_CTRL_III);
+	reg &= ~(SLIH_PAD_DLY_RX1 | SLIH_PAD_DLY_RX0);
+	reg |= FIELD_PREP(SLIH_PAD_DLY_RX1, d1) |
+	       FIELD_PREP(SLIH_PAD_DLY_RX0, d0);
+	mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_III, reg);
+	udelay(1);
+	mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_I,
+		      latch_sel | SLI_CLEAR_BUS | SLI_TRANS_EN | SLI_CLEAR_RX);
+	sli_wait_suspend(SLIH_CPU_BASE);
+	INFO("CPU SLIH[0] US win: {%d, %d} -> select %d\n", d0_start, d0_end,
+	     d0);
+	INFO("CPU SLIH[1] US win: {%d, %d} -> select %d\n", win[d0][0],
+	     win[d0][1], d1);
+}
+
+static int sli_calibrate_mbus_pad_delay(int config, int index, int begin,
+					int end)
+{
+	uint32_t value;
+	uint32_t mask = 0x3f << (index * 6);
+	int d, latch_sel = 0;
+	int win[2] = { -1, -1 };
+
+	if (config)
+		latch_sel = SLI_RX_PHY_LAH_SEL_NEG;
+
+	for (d = begin; d < end; d++) {
+		value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+		value &= ~mask;
+		value |= (d << (index * 6));
+		mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+		udelay(1);
+
+		/* reset IOD TX */
+		value = mmio_read_32(SLIM_IOD_BASE + SLI_CTRL_I);
+		value |= SLI_CLEAR_TX | SLI_CLEAR_BUS;
+		mmio_write_32(SLIM_IOD_BASE + SLI_CTRL_I, value);
+		udelay(1);
+
+		/* Reset CPU RX */
+		mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_I,
+			      latch_sel | SLI_TRANS_EN | SLI_CLEAR_RX |
+				      SLI_CLEAR_BUS);
+		udelay(1);
+
+		/* Check interrupt status */
+		mmio_write_32(SLIM_CPU_BASE + SLI_INTR_STATUS, 0xffffffff);
+		udelay(10);
+		if (is_sli_suspend(SLIM_CPU_BASE) > 0) {
+			if (win[0] == -1)
+				win[0] = d;
+
+			win[1] = d;
+		} else if (win[1] != -1) {
+			break;
+		}
+	}
+
+	if (win[0] == -1)
+		d = SLIM_DEFAULT_DELAY;
+	else
+		d = (win[0] + win[1]) >> 1;
+
+	INFO("CPU SLIM[%d] US win: {%d, %d} -> select %d\n", index, win[0],
+	     win[1], d);
+
+	return d;
+}
+
+static void sli_calibrate_mbus_delay(int config)
+{
+	uint32_t value;
+	int dc, d0, d1, d2, d3;
+	int begin, end;
+	int win[2] = { -1, -1 };
+	int latch_sel = 0;
+
+	if (config)
+		latch_sel = SLI_RX_PHY_LAH_SEL_NEG;
+
+	/* Find coarse delay */
+	for (dc = 0; dc < 16; dc++) {
+		value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+		value &= ~(SLIM_PAD_DLY_RX3 | SLIM_PAD_DLY_RX2 |
+			   SLIM_PAD_DLY_RX1 | SLIM_PAD_DLY_RX0);
+		value |= FIELD_PREP(SLIM_PAD_DLY_RX3, dc) |
+			 FIELD_PREP(SLIM_PAD_DLY_RX2, dc) |
+			 FIELD_PREP(SLIM_PAD_DLY_RX1, dc) |
+			 FIELD_PREP(SLIM_PAD_DLY_RX0, dc);
+		mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+		udelay(1);
+
+		/* reset IOD TX */
+		value = mmio_read_32(SLIM_IOD_BASE + SLI_CTRL_I);
+		value |= SLI_CLEAR_TX | SLI_CLEAR_BUS;
+		mmio_write_32(SLIM_IOD_BASE + SLI_CTRL_I, value);
+		udelay(1);
+
+		/* Reset CPU RX */
+		mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_I,
+			      latch_sel | SLI_TRANS_EN | SLI_CLEAR_RX |
+				      SLI_CLEAR_BUS);
+		udelay(1);
+
+		/* Check interrupt status */
+		mmio_write_32(SLIM_CPU_BASE + SLI_INTR_STATUS, 0xffffffff);
+		udelay(10);
+		if (is_sli_suspend(SLIM_CPU_BASE) > 0) {
+			if (win[0] == -1)
+				win[0] = dc;
+
+			win[1] = dc;
+		} else if (win[1] != -1) {
+			break;
+		}
+	}
+
+	if (win[0] == -1)
+		dc = SLIM_DEFAULT_DELAY;
+	else
+		dc = (win[0] + win[1]) >> 1;
+
+	INFO("CPU SLIM US coarse win: {%d, %d} -> select %d\n", win[0], win[1],
+	     dc);
+
+	value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+	value &= ~(SLIM_PAD_DLY_RX3 | SLIM_PAD_DLY_RX2 | SLIM_PAD_DLY_RX1 |
+		   SLIM_PAD_DLY_RX0);
+	value |= FIELD_PREP(SLIM_PAD_DLY_RX3, dc) |
+		 FIELD_PREP(SLIM_PAD_DLY_RX2, dc) |
+		 FIELD_PREP(SLIM_PAD_DLY_RX1, dc) |
+		 FIELD_PREP(SLIM_PAD_DLY_RX0, dc);
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+	udelay(1);
+
+	begin = (dc - 6) > 0 ? (dc - 6) : 0;
+	end = (dc + 6) < 32 ? (dc + 6) : 32;
+
+	/* Fine-tune per-PAD delay */
+	d0 = sli_calibrate_mbus_pad_delay(config, 0, begin, end);
+	value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+	value &= ~SLIM_PAD_DLY_RX0;
+	value |= FIELD_PREP(SLIM_PAD_DLY_RX0, d0);
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+
+	d1 = sli_calibrate_mbus_pad_delay(config, 1, begin, end);
+	value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+	value &= ~SLIM_PAD_DLY_RX1;
+	value |= FIELD_PREP(SLIM_PAD_DLY_RX1, d1);
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+
+	d2 = sli_calibrate_mbus_pad_delay(config, 2, begin, end);
+	value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+	value &= ~SLIM_PAD_DLY_RX2;
+	value |= FIELD_PREP(SLIM_PAD_DLY_RX2, d2);
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+
+	d3 = sli_calibrate_mbus_pad_delay(config, 3, begin, end);
+	value = mmio_read_32(SLIM_CPU_BASE + SLI_CTRL_III);
+	value &= ~SLIM_PAD_DLY_RX3;
+	value |= FIELD_PREP(SLIM_PAD_DLY_RX3, d3);
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_III, value);
+	udelay(1);
+
+	/* reset IOD SLIM TX */
+	value = mmio_read_32(SLIM_IOD_BASE + SLI_CTRL_I);
+	value |= SLI_CLEAR_TX;
+	mmio_write_32(SLIM_IOD_BASE + SLI_CTRL_I, value);
+	udelay(1);
+
+	/* Reset CPU SLIM Bus (to reset the counters) and RX */
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_I,
+		      latch_sel | SLI_CLEAR_BUS | SLI_TRANS_EN | SLI_CLEAR_RX);
+	sli_wait_suspend(SLIM_CPU_BASE);
+}
+
+static void sli_calibrate_video_delay(int config)
+{
+	uint32_t reg;
+	int d0, d1, d0_start, d0_end;
+	int win[32][2];
+	int i, latch_sel = 0;
+	bool no_win = false;
+
+	if (config)
+		latch_sel = SLI_RX_PHY_LAH_SEL_NEG;
+
+	/* Start SLIV training */
+	for (i = 0; i < 32; i++) {
+		win[i][0] = -1;
+		win[i][1] = -1;
+	}
+
+	/* set max wait count */
+	reg = mmio_read_32(SLIV_CPU_BASE + SLI_CTRL_II);
+	reg |= SLIV_TX_ENT_SUSPEND;
+	mmio_write_32(SLIV_CPU_BASE + SLI_CTRL_II, reg);
+	for (d0 = 0; d0 < 16; d0++) {
+		for (d1 = 0; d1 < 16; d1++) {
+			/* set IOD SLIM DS delay */
+			reg = mmio_read_32(SLIV_IOD_BASE + SLI_CTRL_III);
+			reg &= ~(SLIV_PAD_DLY_RX1 | SLIV_PAD_DLY_RX0);
+			reg |= FIELD_PREP(SLIV_PAD_DLY_RX1, d1) |
+			       FIELD_PREP(SLIV_PAD_DLY_RX0, d0);
+			mmio_write_32(SLIV_IOD_BASE + SLI_CTRL_III, reg);
+			udelay(1);
+
+			/* reset SLIV */
+			mmio_write_32(SLIV_IOD_BASE + SLI_CTRL_I,
+				      latch_sel | 0x8021);
+			mmio_write_32(SLIV_CPU_BASE + SLI_CTRL_I, 0xc021);
+			udelay(200);
+
+			/* check interrupt status */
+			mmio_write_32(SLIV_IOD_BASE + SLI_INTR_STATUS,
+				      0xffffffff);
+			if (is_sli_suspend(SLIV_IOD_BASE) > 0) {
+				if (win[d0][0] == -1)
+					win[d0][0] = d1;
+
+				win[d0][1] = d1;
+			} else if (win[d0][1] != -1) {
+				break;
+			}
+		}
+	}
+
+	d0_start = -1;
+	d0_end = -1;
+	for (i = 0; i < 32; i++) {
+		if (win[i][0] != -1) {
+			if (d0_start == -1)
+				d0_start = i;
+
+			d0_end = i;
+		} else if (d0_end != -1) {
+			break;
+		}
+	}
+
+	if (d0_start == -1) {
+		d0 = SLIV_DEFAULT_DELAY;
+		d1 = SLIV_DEFAULT_DELAY;
+		no_win = true;
+	} else {
+		d0 = (d0_start + d0_end) >> 1;
+		d1 = (win[d0][0] + win[d0][1]) >> 1;
+	}
+
+	reg = mmio_read_32(SLIV_IOD_BASE + SLI_CTRL_III);
+	reg &= ~(SLIV_PAD_DLY_RX1 | SLIV_PAD_DLY_RX0);
+	reg |= FIELD_PREP(SLIV_PAD_DLY_RX1, d1) |
+	       FIELD_PREP(SLIV_PAD_DLY_RX0, d0);
+	mmio_write_32(SLIV_IOD_BASE + SLI_CTRL_III, reg);
+
+	mmio_write_32(SLIV_IOD_BASE + SLI_CTRL_I, latch_sel | 0x8061);
+	mmio_write_32(SLIV_CPU_BASE + SLI_CTRL_I, 0xc061);
+	if (!no_win)
+		sli_wait_suspend(SLIV_IOD_BASE);
+
+	/* clear CPU SLIV interrupt */
+	mmio_write_32(SLIV_IOD_BASE + SLI_INTR_STATUS, 0xffffffff);
+	mmio_write_32(SLIV_CPU_BASE + SLI_INTR_STATUS, 0xffffffff);
+	INFO("IOD SLIV[0] DS win: {%d, %d} -> select %d\n", d0_start, d0_end,
+	     d0);
+	INFO("IOD SLIV[1] DS win: {%d, %d} -> select %d\n", win[d0][0],
+	     win[d0][1], d1);
+}
+
+static void bl31_plat_sli_init(void)
+{
+	uint32_t reg;
+	__maybe_unused const int phyclk_lookup[8] = {
+		25, 800, 400, 200, 2000, 1000, 500, 250,
+	};
+
+	/* early return if SLI had been calibrated */
+	reg = mmio_read_32(SLIH_IOD_BASE + SLI_CTRL_III);
+	reg = FIELD_GET(SLI_PHYCLK_SEL, reg);
+	if (reg)
+		return;
+
+	/* CPU SLI training off */
+	reg = SLI_AUTO_SEND_TRN_OFF | SLI_TRANS_EN;
+	mmio_write_32(SLIM_CPU_BASE + SLI_CTRL_I, reg);
+	mmio_write_32(SLIH_CPU_BASE + SLI_CTRL_I, reg);
+
+	reg |= SLIV_RAW_MODE;
+	mmio_write_32(SLIV_CPU_BASE + SLI_CTRL_I, reg);
+
+	/* change IOD SLIH PHY clock to the target clock frequency */
+	reg = mmio_read_32(SLIH_IOD_BASE + SLI_CTRL_III);
+	reg |= FIELD_PREP(SLI_PHYCLK_SEL, SLI_TARGET_PHYCLK);
+	mmio_write_32(SLIH_IOD_BASE + SLI_CTRL_III, reg);
+	udelay(10);
+
+	if (SLI_TARGET_PHYCLK == SLI_PHYCLK_250M) {
+		sli_calibrate_ahb_delay(1);
+		sli_calibrate_mbus_delay(1);
+	} else {
+		sli_calibrate_ahb_delay(0);
+		sli_calibrate_mbus_delay(0);
+	}
+	sli_calibrate_video_delay(1);
+
+	reg = 0xffffffff;
+	mmio_write_32(SLIH_CPU_BASE + SLI_INTR_STATUS, reg);
+	mmio_write_32(SLIH_IOD_BASE + SLI_INTR_STATUS, reg);
+	mmio_write_32(SLIM_CPU_BASE + SLI_INTR_STATUS, reg);
+	mmio_write_32(SLIM_IOD_BASE + SLI_INTR_STATUS, reg);
+	mmio_write_32(SLIV_CPU_BASE + SLI_INTR_STATUS, reg);
+	mmio_write_32(SLIV_IOD_BASE + SLI_INTR_STATUS, reg);
+
+	INFO("SLI US @ %dMHz init done\n", phyclk_lookup[SLI_TARGET_PHYCLK]);
+}
+
 void bl31_platform_setup(void)
 {
+	generic_delay_timer_init();
+	bl31_plat_sli_init();
 	gicv3_driver_init(&plat_gic_data);
 	gicv3_distif_init();
 	gicv3_rdistif_init(plat_my_core_pos());
diff --git a/plat/aspeed/ast2700/platform.mk b/plat/aspeed/ast2700/platform.mk
index 873c60e20597..f6d86b54de69 100644
--- a/plat/aspeed/ast2700/platform.mk
+++ b/plat/aspeed/ast2700/platform.mk
@@ -18,6 +18,8 @@ BL31_SOURCES += \
 	plat/common/plat_gicv3.c		\
 	plat/common/plat_psci_common.c		\
 	drivers/ti/uart/aarch64/16550_console.S	\
+	drivers/delay_timer/delay_timer.c		\
+	drivers/delay_timer/generic_delay_timer.c	\
 	${PLAT_AST2700}/plat_helpers.S		\
 	${PLAT_AST2700}/plat_topology.c		\
 	${PLAT_AST2700}/plat_bl31_setup.c	\
-- 
2.25.1

